### ISSUE TITLE
Persist properties panel state across sessions

### ISSUE THREAD

**Issue by pinussilvestrus** [camunda/camunda-modeler#2638]
### Is your feature request related to a problem? Please describe.

Currently, when you restart the Modeler(s), the layout state (e.g. groups are collapsed/expanded) got set to the initial state (e.g. all groups are collapsed). The only things we are currently persisting in the `config.json` are the [`width` and the `open` state of the properties container](https://github.com/camunda/camunda-modeler/blob/develop/client/src/app/tabs/PropertiesContainer.js#L192).

```json
"layout": {
  "propertiesPanel": {
    "open": true,
    "width": 250
  }
}
```

![image](https://user-images.githubusercontent.com/9433996/146046405-da366609-7c0d-4255-a158-c4de335a6acc.png)
      
With the new properties panel architecture, we are theoretically able to listen for more specific layout changes as well (cf. https://github.com/bpmn-io/properties-panel-persist-layout-example/pull/1).

### Proposed solution

Assuming that I as a user switch between contexts in order to _compare_ or _copy/paste_:

* [ ] Ensure that properties panel open state is persisted _across_ diagrams.
* [ ] View is restored on a best effort basis; if a section opened does not exist, open it
* [ ] Scroll position is restored on a best effort basis (i.e. open sections) 
* [ ] Store the layout state regarding groups `open` state in the `config.json` and keep it across sessions. As an example, the next time I open the Modeler, the same groups should stay expanded.

As we define a solution for this we should ensure that it works in different environments:

  * [ ] Can be integrated in any place where the properties panel is used
  * [ ] Works in Camunda Modeler
  * [ ] Works in Camunda Web Modeler

### Describe alternatives you've considered

* Instead of storing it globally, store it on a per-file / per-execution platform basis?
* This could be configurable via UI as well
* Not only the layout but also the ordering of groups could be something to be persisted

### Additional context

* https://github.com/bpmn-io/properties-panel-persist-layout-example/pull/1

**Comment by nikku** [comment]
My quick feedback: I'd go for a safe, robust default behavior and not consider the alternatives.

Comparing different diagrams it is of limited use if properties panel layout is a per-file concern; if a particular section is available or the panel has a specific size in one tab, I assume the same holds true in a different tab (per file scope is way to hard to reason about). 

Whether we store it per execution platform depends on how we handle "sections that are expanded", but currently unavailable.

**Comment by andreasgeier** [comment]
The need for this improvement is backed up by [internal feedback](https://docs.google.com/document/d/15FBy6qaziufbIEIpAsHd6hSMy4Coy89QCWrySGiiuQ0/edit#heading=h.ie9wankl1n9o).

Proposed solution:

1. Persist (expanded/collapsed) state of sections across sessions
   1. As a user, I can close a tab and open a new tab while the section states persist
   1. As a user, I can close the application and reopen while the section states persist
1. Sync (expanded/collapsed) state of sections across editor tabs
   1. Example: As a user, I can expand the `Task Definition` section of a `Task` in one editor tab and switch to another editor tab where the section is opened, too (and vice versa)

**Comment by pinussilvestrus** [comment]
I think to achieve both use cases, we should deliver a properties panel API to adjust the layout whenever I want to, e.g.

```js
const eventBus = modeler.get('eventBus');

// update the layout state to initial (all groups closed)
eventBus.fire('propertiesPanel.updateLayout', {
  layout: { groups: {} }
});
```

**Comment by pinussilvestrus** [comment]
**Some observations while working on a spike for this feature**

I timeboxed this feature during spring cleaning (~1 working day).

I pushed an attempt to persist the group layout in the `config.json` [in this branch](https://github.com/camunda/camunda-modeler/tree/2638-persist-layout) (cf. https://github.com/camunda/camunda-modeler/commit/9da329a9b7cd23f5dd8c3a7716926676ca0e7aac)

![Kapture 2022-01-18 at 11 33 38](https://user-images.githubusercontent.com/9433996/149921177-eb6c3bb9-0997-4b0b-8597-75dd59d38d99.gif)

As you see, it kind of works already. What I discovered during implementation
* We have to be safe on committing layout changes (example expand / close groups):
  * change BpmnEditor A + Properties Panel A
  * switch to BpmnEditor B + Properties Panel B 
  * create new BpmnEditor C + Properties Panel C + make changes
  * switch to BpmnEditor B + Properties Panel B
  * => During the whole roundtrip we need to make sure that the layout state between `config.json` all these instances (preact components) is in sync. It's doable, but it needs to be properly tested
* Visually, you already see some flickering of groups when switching around tabs. This is probably a consequence of kind of layout changes flying around. This needs to be observed whether we can limit these events. 
* In comparison to the `minimap` layout, I had [problems while updating the properties panel layout](https://github.com/camunda/camunda-modeler/commit/9da329a9b7cd23f5dd8c3a7716926676ca0e7aac#diff-642e7fb30bb7a37dc4ed7da9086ebfb3be24e2cb853b21a4a423a18646b7b28cR674) since we have a more complex structure here 
* The `async componentDidMount` of the Platform Bpmn Editor seems to make problems (not in the screencast) -> the layout got loaded too late after a tab switch
* Before we make this feature ready, we should decide about our layout structure (and whether it's already what we want). Currently, it looks like

```js
"layout": {
  // ..
  "propertiesPanel": {
    "open": true,
    "width": 250,
    "groups": {
      "general": {
         "open": true
       },
       "documentation": {
         "open": false
       },
       "outputs": {
         "open": false
       },
       "CamundaPlatform__JobExecution": {
         "open": false
       }
     }
   }
}
```

This required some minimal changes in underlying libraries to make sure we properly commit layout changes
* https://github.com/bpmn-io/bpmn-js-properties-panel/tree/2638-update-layout-api
* https://github.com/bpmn-io/properties-panel/tree/2638-set-layout-api
* https://github.com/bpmn-io/properties-panel-persist-layout-example/tree/2638-set-layout-api (Example)

**Summary**: To make this feature properly work, we need some extra time to make the layout persisting robust (e.g. making sure we don't fire unnecessary `propertiesPanel.layoutChanged` events, investigate why it lacks with `async componentDidMount`).

**Comment by pinussilvestrus** [comment]
To be considered: whatever we come up as an improvement [in the lining department](https://github.com/bpmn-io/internal-docs/issues/471) could be re-used in this topic.

/cc @philippfromme

**Comment by nikku** [comment]
@christian-konrad @CatalinaMoisuc I propose that we pick this up as the next bigger topic in core editing. The proposed solution is to restore the properties-panel@0.x behavior. It has been been validated numerous times in a number of bug reports raised folks duplicating this bug). 

The current situation is a huge UX blunder.

**Comment by nikku** [comment]
Clarified that we aim for a solution that is generally applicable to all environments where the properties panel embeds.

**Comment by marstamm** [comment]
As discussed on [slack](https://camunda.slack.com/archives/GP70M0J6M/p1677590033828309), we will have 2 different behaviors for Web and Desktop applications.

- Web: Opening/Closing Sections should be persisted for new tabs/reloads. It does not affect other opened tabs, the latest state will be persisted.
- Desktop: Opening/Closing Sections also affects the opened Sections in other, already opened tabs. E.g., opening the "General" Group will result in the group being opened in other tabs when we switch to them, even when they were previously closed.

## Breakdown

### @bpmn-io/properties-panel

- Groups: don't use local state in [`useLayoutState`](https://github.com/bpmn-io/properties-panel/blob/e01dd051aba1b4998bcc6f7a591ad7cdf8fc4cf7/src/hooks/useLayoutState.js#L26). Instead, always use the Context value, so outside changes in the layout are reflected in the Groups.
- PropertiesPanel (root): React to `layoutConfig` changes, e.g. from other tabs. Currently, only the initial value is used.

### bpmn-js-properties-panel

- listen to `propertiesPanel.layoutChanged` events and update the layoutConfig supplied to `@bpmn-io/properties-panel` accordingly
- Persist layout in localStorage and supply it to the properties-panel

### Desktop Modeler

- listen to `propertiesPanel.layoutChanged` and update global layout state (this state already exists for property panel width)
- for cached modelers, fire `propertiesPanel.layoutChanged` on component mount to update the cached property panel layout with the global layout
